### PR TITLE
Swift4 Fix #48

### LINKED ISSUE #48: Support Swift 4
When do you plan to add support for Swift 4

### PR DESCRIPTION
**Pull request by Ponyboy47**
1) Removed unnecessary imports
2) Removed _FMWrapper class and replaced with the default FileManager
3) Implemented subscripting using new Partial Ranges
4) Re-Implemented `|>>` using the TextFileStreamWriter
5) Replaced NSString and NSData in places where String or Data would also work
6) Added compiler flags for macOS as well as OSX, because I suspect the OSX flag will be deprecated at some point
7) Minor work towards including Linux support
8) Removed the _FMWrapper tests
9) Updated swift version and FileKit version

### CODE REVIEW

**Comment by nvzqz** [comment]
Awesome! I'll take a thorough look at this tonight.

**Comment by Ponyboy47** [comment]
Sorry that I keep adding commits. I've been slowly refactoring stuff as I go

**Comment by nvzqz** [comment]
@Ponyboy47 please see my comment regarding https://github.com/nvzqz/FileKit/pull/49/commits/fffec4d9c9565bc713520f366a5bba30dc3ea249#diff-fc6e29b0e81d4e4167956692fa94b444R938.

Xcode 9 is in its third beta. When Xcode 9 hits GM or the macOS App Store, then the Swift version project setting should change to accommodate the release.

File: Sources/ArrayFile.swift [hunk 1]
```diff
@@ -27,12 +27,16 @@
...
 
-/// A representation of a filesystem array file.
-///
-/// The data type is NSArray.
+/**
+ A representation of a filesystem array file.
+
+ The data type is NSArray.
+*/
 public typealias NSArrayFile = File<NSArray>
```
Comment (phimage) [inline Sources/ArrayFile.swift]: Is it a swift4 needs to change all the comments?
If not, a separated PR would have been better (to see only swift 4 change, and discuss elsewhere about comment formatting)

File: Sources/NSString+FileKit.swift [hunk 2]
```diff
@@ -38,28 +38,30 @@ extension NSString {
...
+                             the file directly.
+    */
     public func write(to path: Path, atomically useAuxiliaryFile: Bool) throws {
-        do {
-            try self.write(toFile: path._safeRawValue,
+        guard let _ = try? self.write(toFile: path._safeRawValue,
                            atomically: useAuxiliaryFile,
                            encoding: String.Encoding.utf8.rawValue)
-        } catch {
+        else {
```
Comment (phimage) [inline Sources/NSString+FileKit.swift]: This code do not allow to debug using a breakpoint to see the error.
Swift4 is a breaking change so I think `FileKitError.writeToFileFail` must contain the error as second argument

File: Sources/Path.swift [hunk 3]
```diff
@@ -47,66 +48,44 @@ public struct Path {
...
     public static func volumes(_ options: FileManager.VolumeEnumerationOptions = []) -> [Path] {
-        let volumes = FileManager.default.mountedVolumeURLs(includingResourceValuesForKeys: nil,
+        let volumes = Path.fileManager.mountedVolumeURLs(includingResourceValuesForKeys: nil,
             options: options)
         return (volumes ?? []).flatMap { Path(url: $0) }
     }
 
     // MARK: - Properties
 
-    fileprivate var _fmWraper = _FMWrapper()
```
Comment (phimage) [inline Sources/Path.swift]: Using a custom FileManager is a useful macOS feature, not really used for iOS
When you have a lot to do with files, you want to separate your `FileManagerDelegate` codes

File: Sources/Path.swift [hunk 4]
```diff
@@ -119,38 +98,42 @@ public struct Path {
...
-    ///
-    /// Return [] if path is `.` or `""`
+    /**
+     The components of the path.
+
+     Returns [] if path is `.` or `""`
+    */
     public var components: [Path] {
-        if rawValue == "" || rawValue == "." {
+        guard rawValue != "" && rawValue != "." else {
```
Comment (phimage) [inline Sources/Path.swift]: `guard` is not always the solution

I see a lot `if` replaced by `guard` with tricky boolean conditions

it's a matter of coding style, `if` has been clear here, `guard` also is clear

File: Sources/TextFile.swift [hunk 5]
```diff
@@ -274,27 +293,34 @@ open class TextFileStreamWriter: TextFileStream {
...
-            handle.write(data)
-            if delim {
-                handle.write(delimData)
-            }
-            return true
+        guard let handle = fileHandle, let data = line.data(using: self.encoding) else {
+            return false
+        }
+        if delim && append {
+            handle.write(delimData)
```
Comment (phimage) [inline Sources/TextFile.swift]: This is a behaviour change
if append mode, you want the delim only before?
This change must be documented carefully(method comment, changelog when releasing) or be reverted

File: Sources/ArrayFile.swift (hunk 1 above)
Comment (nvzqz) [reply Sources/ArrayFile.swift]: Swift 4 _does not_ require a change in comment style. I've updated [RandomKit](https://github.com/nvzqz/RandomKit) to Swift 4 and much has remained the same as before, including the comment style.

**Comment by phimage** [comment]
Thanks @Ponyboy47. As you can see your PR could not be merged.
I convert myself to not have all comments modified.

I do some more work because of your PR like "Re-Implemented |>> using the TextFileStreamWriter"
I do not do removing import or linux stuff.
You can PR for that later

**Pull request closed without merging**